### PR TITLE
Add Dockerfile

### PR DESCRIPTION
**Pull request by tmaier**
Run gitlab-le within a docker container.

After merging this, set up a docker repository like this one: https://hub.docker.com/r/tmaier/gitlab-letsencrypt/

- login at https://hub.docker.com
- "Create Automated Build"
- Configure "Build Settings" --> "Repository Links". Add here `node`

Use the following command:

```shell
docker container run --rm -it tmaier/gitlab-letsencrypt --domain example.com --email me@example.com --repository https://gitlab.com/my/repo --jekyll --path /acme-challenge --token $MY_TOKEN
```

Schedule a job with GitLab CI:

Note: you need to set `GITLAB_TOKEN` as variable and you need to enable the [Pipeline Schedule](https://docs.gitlab.com/ce/user/project/pipelines/schedules.html)

```yaml
ssl:renew certificate:
  image:
    name: tmaier/gitlab-letsencrypt
    entrypoint: ["/bin/sh", "-c"]
  stage: deploy
  before_script: []
  script: |-
    gitlab-le \
      --domain example.com \
      --email me@example.com \
      --jekyll \
      --path /acme-challenge \
      --production \
      --repository $CI_PROJECT_URL \
      --token $GITLAB_TOKEN
  only:
  - schedules
```

### CODE REVIEW

**Comment by tmaier** [comment]
The failure of Travis CI is not related to this Pull Request

**Comment by axilleas** [comment]
@tmaier that's great! I followed a similar approach in my `.gitlab-ci.yml` :) The email can also be added as a secret variable to avoid spam if the project is public.

Can you also add a couple of words in the readme? If this is merged as is, people will not know how to use the Docker image :)

**Comment by tmaier** [comment]
Hi @axilleas,
you are right. I am updating the README as we speak.

Do you know if there are any constraints on the E-Mail-Adress to use for Let's Encrypt?
If there are none, I would go for `GITLAB_USER_EMAIL` (see https://docs.gitlab.com/ce/ci/variables/README.html)

**Comment by tmaier** [comment]
Please review the updated README at https://github.com/tmaier/gitlab-letsencrypt/blob/master/README.markdown

**Comment by tmaier** [comment]
The latest commit reduces the docker image size from 92 MB to 20 MB.

File: README.markdown [hunk 1]
```diff
@@ -45,6 +46,53 @@ Try it out: https://example.com https://www.example.com (GitLab might take a few
...
+      --production \
+      --repository $CI_PROJECT_URL \
+      --token $GITLAB_TOKEN
+  only:
+  - schedules
+```
+
+Add the following variables to your GitLab project: `LETS_ENCRYPT_EMAIL` and your secret `GITLAB_TOKEN`.
+
+Schedule then a new pipeline to run for example every month. See <https://docs.gitlab.com/ce/user/project/pipelines/schedules.html> for details
```
Comment (axilleas) [inline README.markdown]: Missing period :)

File: README.markdown [hunk 2]
```diff
@@ -45,6 +46,53 @@ Try it out: https://example.com https://www.example.com (GitLab might take a few
 This certificate expires on Sat Apr 14 2018 03:09:06 GMT+0100 (BST). You will need to run gitlab-le again at some time before this date.
 ```
 
+## Docker image
+
+There is also a [Docker image](https://hub.docker.com/r/rolodato/gitlab-letsencrypt/) available.
+This means the Command-line tool can be used without installing all the dependencies required to run the application.
+
+Example:
```
Comment (axilleas) [inline README.markdown:54]: Can you leave a newline here? :)

**Review by axilleas — COMMENTED** [review]
@tmaier that's great! I left a couple of comments.

File: README.markdown [hunk 3]
```diff
@@ -45,6 +46,53 @@ Try it out: https://example.com https://www.example.com (GitLab might take a few
...
+They need to be renewed regularly.
+
+Use the [GitLab Pipeline Schedule](https://docs.gitlab.com/ce/user/project/pipelines/schedules.html) feature to automate the renewal process.
+
+```yaml
+ssl:renew certificate:
+  image:
+    name: rolodato/gitlab-letsencrypt
+    entrypoint: ["/bin/sh", "-c"]
+  stage: deploy
```
Comment (axilleas) [inline README.markdown]: This should be run after the site build right? Maybe add a note about the stage naming.

**Comment by axilleas** [comment]
> Do you know if there are any constraints on the E-Mail-Adress to use for Let's Encrypt?
If there are none, I would go for GITLAB_USER_EMAIL

I'm not aware of anything about this. I guess we can use this variable :)

File: README.markdown (hunk 3 above)
Comment (tmaier) [reply README.markdown]: It does not really matter in which stage it runs. I would recommend to add to all other jobs `except: [schedules]`, as the next build will be triggered by gitlab-le anyway

File: README.markdown (hunk 2 above)
Comment (tmaier) [reply README.markdown:54]: What do you mean? I think there is a newline due to the code block following this line

**Comment by rolodato** [comment]
Hey, thanks for putting this together! Seems like a great way to solve the automation problem. I'm not sure I'd like to maintain this though, since I am not very familiar with Docker.

If you don't mind, I'd rather keep this on a separate repository (your fork or any other repo would be fine) and I can link to it from the Setup and Automation section of the README. That way it's still available for anyone who is interested. What do you think?

**Comment by tmaier** [comment]
Hi @rolodato,

normally, I would like to keep it within this repo, as for me Docker support and this kind of integration with GitLab CI belong directly to this repo and the scope of the tool.

But I understand your concern.

If I remove the Dockerfile, shall I keep the changes to the README in this form or another?

Under which cirumstances would you be ready to have the Dockerfile inside this repository? For example, do the maintainer labels in the Dockerfile help? Or if there would be contribution guidelines which states that Docker is not supported and that questions on Docker+gitlab-le shall be placed on StackOverflow?

**Comment by rolodato** [comment]
Hi @tmaier, sorry for the delay getting back to you.

> If I remove the Dockerfile, shall I keep the changes to the README in this form or another?

I would just leave a link to the Dockerfile in the Automation section, wherever it ends up being hosted (your repo or anywhere else is fine).

> Under which cirumstances would you be ready to have the Dockerfile inside this repository? For example, do the maintainer labels in the Dockerfile help? Or if there would be contribution guidelines which states that Docker is not supported and that questions on Docker+gitlab-le shall be placed on StackOverflow?

As I mentioned in my previous comment, since I'm not very familiar with Docker in general, I would prefer to leave the Dockerfile for others to maintain. Even if Dockerfiles have some way of defining external maintainers, I would still need to manage pull requests and issues for the Dockerfile if it ends up on this repository.

In short, please publish the Dockerfile yourself, with a README or other instructions if you'd like, and I'd be happy to link them from this repo's README.

Thanks!
